packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart

```diff
@@ -0,0 +1,173 @@
+///      if the caller sets the completer on the stream.
+///   1. If the [ImageCache] has a completer for the key for this image, ask the
+///      wrapped provider to resolve.
+///      This can happen if the image was precached, or another [ImageProvider]
+///      already resolved the same image.
+///   1. If the [context] has been disposed, end. This can happen if the caller
+///      has been disposed and is no longer interested in resolving the image.
+///   1. If the widget is scrolling with high velocity at this point in time,
+///      wait until the beginning of the next frame and go back to step 1.
+///   1. Delegate loading the image to the wrapped provider and finish.
+///
+/// The [Image] widget wraps its incoming providers with this provider to avoid
+/// overutilization of resources for images that would never appear on screen or
+/// only be visible for a very brief period.
+@optionalTypeArgs
+class ScrollAwareImageProvider<T> extends ImageProvider<T> {
+  /// Creates a [ScrollingAwareImageProvider].
+  ///
+  /// The [context] object is the [BuildContext] of the [State] using this
+  /// provider. It is used to determine scrolling velocity during [resolve]. It
+  /// must not be null.
+  ///
+  /// The [imageProvider] is used to create a key and load the image. It must
+  /// not be null, and is assumed to interact with the cache in the normal way
+  /// that [ImageProvider.useKey] does.
+  const ScrollAwareImageProvider({
+    @required this.context,
+    @required this.imageProvider,
+  }) : assert(context != null),
+       assert(imageProvider != null);
+
+  /// The context that may or may not be enclosed by a scrollable.
+  ///
+  /// Once [ScrollAwareContextProvider.dispose] is called on this context,
+  /// the provider will
+  final ScrollAwareContextProvider context;
+
+  /// The wrapped image provider to delegate [obtainKey] and [load] to.
+  final ImageProvider<T> imageProvider;
+
+  @override
+  void useKey(
+    ImageConfiguration configuration,
+    ImageStream stream,
+    T key,
+    ImageErrorListener handleError,
+  ) {
+    assert(stream != null);
+    assert(key != null);
+    assert(handleError != null);
+
+    void deferredResolve([bool firstCall = false]) {
+      // Something managed to complete the stream. Nothing left to do.
+      if (stream.completer != null) {
+        return;
+      }
+      // Something else got this image into the cache. Return it.
+      if (PaintingBinding.instance.imageCache.containsKey(key)) {
+        imageProvider.useKey(configuration, stream, key, handleError);
+      }
+      // The context has gone out of the tree - ignore it.
+      if (context._context == null) {
+        return;
+      }
+      // Something still wants this image, but check if the context is scrolling
+      // too fast before scheduling work that might never show on screen.
+      // Try to get to end of the frame callbacks of the next frame, and then
```

Comment:
Why try to? What could stop you? (Other then the engine not requesting any more frames?

Reply:
Someone else could call `scheduleFrameCallback` after us and then schedule a microtask, or the engine could stop requesting frames.